### PR TITLE
vdk-control-api-auth: Add core auth logic

### PR DESCRIPTION
**Pull request by doks5**
As part of the work on extracting the authentication logic of vdk-control-cli
and moving it into a stand-alone library, we need to have a generalized
approach to authentication which is compliant with RFC-6749 while being able
to be integrated with the existing components of Versatile Data Kit, which
require authentication.

This change introduces the core authentication logic, which will be used and
extended to facilitate the different authentication workflows (api/refresh token
flow, pkce flow, etc.).

This is the first Pull Request of a series that will build the complete authentication library (plugin). The
plugin will not be released before all work is done.

Testing Done: Unit tests added.

Signed-off-by: Andon Andonov <andonova@vmware.com>

### CODE REVIEW

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_config.py
Comment (gageorgiev) [inline projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_config.py:110]:
```suggestion
                    raise VDKAuthException(
                        what="Configuration folder was not created and user is not logged in",
                        why=f"Cannot create: {self.vdk_config_folder} configuration folder: {str(e)}",
                        consequence="User won't be able to access Control Service",
                        countermeasure=f"Check if the client has write access to: {base_dir}",
                    ) from e
```
Is including the exception string inside the why parameter just as good as raising from the original exception? It might be, and if so you can ignore this suggestion.

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py
Comment (gageorgiev) [inline projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py:194]: nit but maybe move __exchange_refresh_for_access_token up so it's next to __exchange_api_for_access_token

**Review by gageorgiev — APPROVED** [review]
Looks solid; it's a big PR so maybe wait for an additional approve before merging in case I've missed something.

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py]: Moved. Thanks

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_config.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_config.py:110]: I am hesitant to excessively clutter the logs, but if others think the whole exception should be logged, as well, I am fine to go with the suggested change.

File: projects/vdk-plugins/vdk-control-api-auth/setup.py
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-control-api-auth/setup.py:28]: why not reusing vdk-core's configuration mechanism?

File: projects/vdk-plugins/vdk-control-api-auth/setup.py
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-control-api-auth/setup.py:31]: looking at vdk-control-api-auth README.md, is it meant to be used by other plugins like "library";
maybe imported & explicitly instantiated? like common reusable code? is it concurrent? it is not clear by the README, so I'm just guessing; maybe provide an example snippet in the readme?

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_utils.py
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_utils.py:13]: what's the intended usage? didn't find any usages (except in test utils, where a duplicate's in usage )

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py:6]: maybe mark with #deprecated the duplicates (e.g. `auth_request_values.py` in vdk-control-cli's auth package), so it's easier to diff current code with prior, and to locate what needs to be cleaned up afterwards?

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py:18]: you mention pkce within the PR description, and I don't find any such related changes?
so `auth_pkce.py` from vdk-control-cli remains?

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py:8]: are we extracting also RedirectAuthentication for reusing?

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py:35]: why None in favour of priorly used string placeholder by `auth.py`?

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py:94]: why `BaseAuth` in favour of the old `Authentication`? is it abstract and meant for extension? don't see any prior extensions;
also, transitioning the imports only is going to be easier

File: projects/vdk-plugins/vdk-control-api-auth/tests/test_core_auth.py
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-control-api-auth/tests/test_core_auth.py:3]: why file named core? is there something in addition to test?

File: projects/vdk-plugins/vdk-control-api-auth/tests/test_core_auth.py
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-control-api-auth/tests/test_core_auth.py:218]: why not reusing test utils?

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py:9]: Do you think it's a good idea to transfer also `ApiKeyAuthentication`, since it does make use of the Authentication class to instantiate it?
It contains the usage workflow; is it concurrent?

**Review by ivakoleva — COMMENTED** [review]
Since the library plugin is meant to be used by other plugins, maybe add functional tests that do exactly that and verify operability?
for example, access (read) the freshest access token generated;

If I recall correctly, based on the provider specifics - old tokens may be invalidated (after a certain period of time, or even immediately) - on token renewal; 
so multiprocessing for scalability has to do with concurrent Authentication support.

File: projects/vdk-plugins/vdk-control-api-auth/setup.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/setup.py:28]: To not rely on vdk-core and its dependencies. I am hoping for the library to be as light as possible, as we will use it for the airflow provider.

File: projects/vdk-plugins/vdk-control-api-auth/setup.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/setup.py:31]: Yes, it is a library, I'll update the README with subsequent PRs. This is only the first PR, and currently the package is not yet being released.
It is not currently concurrent, do we have a requirement for libraries to be concurrent?

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_exception.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_exception.py:5]: The vdk-control-cli's VdkException extends `click.UsageError` in order to allow click to properly display the error message in terminal.
I did not re-use the exceptions from vdk-core/vdk-control-cli to avoid circular dependencies.

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_utils.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_utils.py]: Oopsie, good catch. This one was supposed to be part of a separate PR. Removed.

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py:6]: Will do as part of a separate PR.

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py:18]: Just updated the PR description. The pkce authentication flow will be added as part of a separate PR. I didn't want to put everything in a single PR, as the change would be really hard to review properly, so I opted for a multi-PR change.
The release step in the CI/CD will not be enabled until the work on the library is complete.

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py:8]: Yes, we will. See https://github.com/vmware/versatile-data-kit/pull/815#discussion_r860761342

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py:35]: To get rid of the codacy warning messages.

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/base_auth.py:94]: There will be a different `Authentication` api with more generalized logic.

File: projects/vdk-plugins/vdk-control-api-auth/tests/test_core_auth.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/tests/test_core_auth.py:3]: Yes, there will be additional tests for the different authentication flows.

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_request_values.py:9]: Yes, `ApiKeyAuthentication` will also be transferred, alongside the other authentication flows, as separate PRs to make the changes easier to review.

File: projects/vdk-plugins/vdk-control-api-auth/setup.py
Comment (ivakoleva) [reply projects/vdk-plugins/vdk-control-api-auth/setup.py:28]: the airflow provider is not vdk-core-dependant?

File: projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_exception.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/src/vdk/plugin/control_api_auth/auth_exception.py:5]: Not exactly. There will be a few more PRs for the different authentication flows. Then I'll enable the release step of the librry, and only after that there will be a PR with the vdk-control-cli integration with this library.

File: projects/vdk-plugins/vdk-control-api-auth/setup.py
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/setup.py:28]: Not currently, https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-plugins/airflow-provider-vdk/setup.py#L28
Not sure if there are plans to have vdk-core as a dependency.

File: projects/vdk-plugins/vdk-control-api-auth/.plugin-ci.yml
Comment (tozka) [inline projects/vdk-plugins/vdk-control-api-auth/.plugin-ci.yml:12]: Why not reuse .build-plugin from plugin-common ?

**Review by tozka — APPROVED** [review]

File: projects/vdk-plugins/vdk-control-api-auth/requirements.txt
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/requirements.txt:4]: Good catch. Will do

File: projects/vdk-plugins/vdk-control-api-auth/.plugin-ci.yml
Comment (doks5) [reply projects/vdk-plugins/vdk-control-api-auth/.plugin-ci.yml:6]: @tozka Updated the main plugin CI/CD script to ignore the entry-point of the plugin, similar to how `quickstart-vdk` and `vdk-test-utils` are handled. We may need to think about a different way to handle such plugins, though, as this does not seem manageable in the long run.

Comment (tozka) [reply projects/vdk-plugins/vdk-control-api-auth/.plugin-ci.yml:6]: Yes, I agree, it's not ideal. 2 options come to mind: 

A) We remove the check. If we think what the check is doing (testing the "wiring") is covered by other tests . 
B) Make the check option but reverse - that is expose variable (CHECK_VDK_PLUGIN_ENTRYPOINT=false/true) and those 3 plugins/libs would set it to false.

**Pull request merged**
